File: 233.number-of-digit-one.py
#
# @lc app=leetcode id=233 lang=python3
#
# [233] Number of Digit One
#

# @lc code=start
class Solution:
    def countDigitOne(self, n: int) -> int:
        counts = [1,]
        res = 0
        bits = [int(i) for i in list(str(n))]
        count_map = [0]
        for i in  range(1,len(bits)):
            count_map.append(10**(i-1)+10*count_map[-1])
            # print(count_map)
            # count_map.append(10**(i-1)+9*count_map[-1])
        # count_map.pop(0)
        # count_map = [sum(count_map[:i+1]) for i in range(len(count_map))]
        count_map.reverse()
        # print(count_map)
        print(count_map)
        for (index,bit) in enumerate(bits):
            res += bit * count_map[index]
            if bit > 1:
                pass
            elif bit == 1:
                pass
            else:
                pass
        return res

        
        
# @lc code=end
if __name__ == "__main__":
    print(Solution().countDigitOne(54321))

File: golang/233.number-of-digit-one.go
/*
 * @lc app=leetcode id=233 lang=golang
 *
 * [233] Number of Digit One
 */
package main

//  digit dynamic programming is naturally suitable for this problem
// @lc code=start
var (
	digits []int{}
	dp 	[]int{}
)
func countDigitOne(n int) int {
	for n > 0 {
		digits = append([]int{n % 10}, digits...)
		n %= 10
	}
	pos:=0
}

func dfs(digits,pos,dp){
	ret:=0
	for i:=0;i<len(digits);i++{
		for j:=0;j<digits[i];j++{
			ret+=dfs(digits,n+1)
		}
	}

}
// @lc code=end
func main() {
	// fmt.Printf("%d", countDigitOne(13))
}
